In the REPL show extended explanations only once

som-snytt opened the pull request:
Fixes scala/bug#10943

Per-run reporting adds an extended explanation only once per feature warning.

REPL wants to warn once per resident session.

This PR hooks `featureWarning` to avoid multiple explanations, and also avoids repeated "extended explanations" in the reporter.

som-snytt committed:
REPL shows boring explanations only once

dwijnand commented:
(Couldn't tell if it did already and it shouldn't or the other way round, so I tweaked the description.)

som-snytt committed:
REPL intercepts feature warning

som-snytt commented:
Updated to intercept the feature warning directly, where the REPL global remembers the "feature names" it has already warned about. (PerRunReporting remembers Symbols.)

Simplified the reporter-side processing, and left it operational for non-feature messaging with an "extended explanation". I don't know if there are any such messages currently.

dwijnand commented:
What's the difference between the symbol and the name?  Is the featureTrait symbol not the symbol for the feature and as unique as its name?

dwijnand commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -199,6 +199,8 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
     }
 
     private[this] var reportedFeature = Set[Symbol]()
+    protected def featureReported(featureTrait: Symbol): Unit = reportedFeature += featureTrait
```
Reuse in `featureWarning`?

som-snytt replied:
I was lazy.

som-snytt commented:
I didn't at-ask Lukas whether symbol was more specific than name, but I thought it's safer holding onto a string for the life of the global.
